man/formhame.1
.TH  Tukubai 1 "28 Jun 2022" "usp Tukubai" "Tukubai コマンド マニュアル"

.br
formhame(1)
.br

.br
【名前】
.br

.br
formhame\ :\ HTML\ テンプレートに文字をはめ込む
.br

.br
【書式】
.br

.br
Usage\ \ \ :\ formhame\ <html_template>\ <data>
.br
Option\ \ :\ -i<c>
.br
        : -d<c>
.br
Version\ :\ Sun\ Mar\ 17\ 12:51:52\ JST\ 2013
.br
Edition\ :\ 1
.br

.br
【説明】
.br

.br
<html_template>ファイルの各種inputタグ(text,\ radio,checkbox,\ hidden)とtextarea
.br
タグ、selectタグの部分に<data>ファイル(第1フィールド:タグ名、第2フィールド以降:
.br
値)で指定される値を挿入する。
.br

.br
【例】
.br

.br

  $ cat html
  <html><body>
  <form name="name_form">
  <input type="text" name="name_text1" />
  <input type="text" name="name_text2" />
  <input type="radio" name="name_radio" value="a"/>
  <input type="radio" name="name_radio" value="b"/>
  <input type="checkbox" name="name_checkbox" value="x"/>
  <input type="checkbox" name="name_checkbox" value="y"/>
  <textarea name="name_textarea">
  </textarea>
  <select name="name_pulldown">
  <option value="pd1">pd1</option>
  <option value="pd2">pd2</option>
  <option value="pd3">pd3</option>
  </select>
  <input type="submit" name="submit" />
  </form>
  </body></html>

  $

.br

  $ cat data
  name_text1 hello
  name_text2
  name_radio b
  name_checkbox y
  name_textarea usp\nlaboratory
  name_pulldown pd3

  $

.br

  $ formhame html data
  <html><body>
  <form name="name_form">
  <input type="text" name="name_text1" value="hello"/>
  <input type="text" name="name_text2" />
  <input type="radio" name="name_radio" value="a"/>
  <input type="radio" name="name_radio" value="b" checked="checked" />
  <input type="checkbox" name="name_checkbox" value="x"/>
  <input type="checkbox" name="name_checkbox" value="y" checked="checked" />
  <textarea name="name_textarea">
  usp
  laboratory
  </textarea>
  <select name="name_pulldown">
  <option value="pd1">pd1</option>
  <option value="pd2">pd2</option>
  <option value="pd3" selected="selected">pd3</option>
  </select>
  <input type="submit" name="submit" />
  </form>
  </body></html>

  $

.br
【備考】
.br

.br
\ 1.\ inputタグの場合、すでにvalue=""とテンプレートに記述している場合は、値を置換
.br
\ \ \ \ する。同様にtextareaタグの場合も値を置換する。selectタグの場合は、すでに
.br
\ \ \ \ selected="selected"とテンプレートに指定されている場合、指定場所を移動する。
.br
\ 2.\ textareaの場合、すでにある値を置換する。.B\ \e
nは改行に置換する。
.br
\ 3.\ 値が無い場合、値は挿入されない。
.br
\ 4.\ -iオプションで文字列を指定すると、その文字列に等しい値はヌル文字列にして挿
.br
\ \ \ \ 入する。
.br
\ 5.\ -dオプションで文字列を指定すると、値の中の指定文字列部分は半角空白文字に置換して
.br
\ \ \ \ 挿入する。ただし、.B\ \e
指定文字列という部分は半角空白ではなく（.B\ \e
をとった）指定文字列に置換して挿入する。
.br
\ 6.\ -iオプションと-dオプションの両方が指定された場合、-iオプションによる処理を先に行なう。
.br
\ 7.\ HTMLは正しく記述されている必要がある。タグは<タグ\ />あるいは、<タグ></タグ>と記述される必要がある。
.br
\ \ \ \ 変数は必ず"\ "で囲まれている必要がある(value="1"など)。
.br
\ \ \ \ input\ type="checkbox"のときは、必ずvalue="XXX"が指
.br
\ \ \ \ 定されている必要がある。
.br
\ 8.\ checkbox、radiobox、selectboxの複数選択の場合は、ネーム形式データ(5)は次の
.br
\ \ \ \ ように1つのタグに対して複行となる。
.br

.br
name_radio\ a
.br
name_radio\ b
.br
name_checkbox\ x
.br
name_checkbox\ y
.br
name_pulldown\ pd1
.br
name_pulldown\ pd2
.br

.br
【関連項目】
.br

.br
filehame(1)、mojihame(1)
.br

.br
last\ modified:\ Mon\ Aug\ \ 1\ 16:43:39\ JST\ 2022
.br
Contact\ us:\ uecinfo@usp-lab.com
.br
Copyright\ (c)\ 2012-2022\ Universal\ Shell\ Programming\ Laboratory\ All\ Rights
.br


COMMANDS.HS/formhame.hs
#!/usr/bin/env runghc
import System.Environment
import System.IO
import Text.ParserCombinators.Parsec
import Control.Monad
import Control.Applicative hiding ((<|>), many)
import Data.List 
import qualified Data.Text as DT (replace)
import qualified Text.Regex.Posix as RE

{--
formhame（Open usp Tukubai）

designed by Nobuaki Tounaka
written  by Hinata Yanagi

The MIT License

Copyright (C) 2022 Universal Shell Programming Laboratory

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
--}

showUsage :: IO ()
showUsage = do
    System.IO.hPutStr stderr "Usage    : formhame <html_template> <data>\n"
    System.IO.hPutStr stderr "Version  : Sat Oct  1 21:43:34 JST 2022\n"
    System.IO.hPutStr stderr "Open usp Tukubai (LINUX+FREEBSD)\n"

data Opts = NormalOpts TemplateFile DataFile | Error String deriving Show

type KeyValue = (String,String)

type TemplateFile = String
type DataFile = String

main :: IO ()
main = do args <- getArgs
          case args of
              []                       -> showUsage
              ["-h"]                   -> showUsage
              ["--help"]               -> showUsage
              ["--version"]            -> showUsage
              [('-':'i':istr),('-':'d':dstr),f1,f2]    -> main' istr dstr f1 f2
              [('-':'i':istr),('-':'d':dstr),f1]    -> main' istr dstr f1 "-"
              [('-':'d':dstr),('-':'i':istr),f1,f2]    -> main' istr dstr f1 f2
              [('-':'d':dstr),('-':'i':istr),f1]    -> main' istr dstr f1 "-"
              [('-':'i':istr),f1,f2]    -> main' istr "" f1 f2
              [('-':'i':istr),f1]    -> main' istr "" f1 "-"
              [('-':'d':dstr),f1,f2]    -> main' "" dstr f1 f2
              [('-':'d':dstr),f1]    -> main' "" dstr f1 "-"
              [f1,f2]                  -> main' "" "" f1 f2
              [f1]                     -> main' "" "" f1 "-"

main' :: String-> String -> String -> String -> IO ()
main' istr dstr tempfile datafile = do tempcs <- readF tempfile
                                       datacs <- readF datafile
                                       formhame ( readData istr dstr (lines datacs) ) (lines tempcs)

readF :: String -> IO String
readF "-" = getContents
readF f = readFile f

readData :: String -> String -> [String] -> [KeyValue]
readData istr dstr lns = map ( (fd dstr) . (fi istr) . f . words) lns
    where f (a:b:c) = (a,b)
          f (a:[])  = (a,"")
          fi istr (a,b) = if istr == b then (a,"") else (a,b)
          fd dstr (a,b) = (a,sub dstr b)

sub :: String -> String -> String
sub ""   str = str
sub dstr str
 | length dstr > length str     = str
 | ('\\':dstr) `isPrefixOf` str = dstr ++ (sub dstr $ drop (1 + length dstr) str )
 | dstr `isPrefixOf` str        = " " ++ (sub dstr $ drop (length dstr) str )
 | otherwise                    = (take 1 str ) ++ sub dstr (drop 1 str)

formhame :: [KeyValue] -> [String] -> IO ()
formhame _ [] = return ()
formhame kvs (ln:lns) 
  | pickedkvs == [] = putStrLn ln >> formhame kvs lns
  | otherwise       = replaceValue tp pickedkvs (fst x) >> formhame kvs (snd x)
    where pickedkvs = pickKvs kvs ln
          x = cutTargetHtml tp (ln:lns)
          tp = checkType ln

replaceValue :: String -> [KeyValue] -> [String] -> IO ()
replaceValue "text"  (kv:[]) (ln:[]) = replaceValueText kv ln
replaceValue "radio" (kv:[]) (ln:[]) = replaceValueRadio kv ln
replaceValue "checkbox" kvs  (ln:[]) = replaceValueChkbox kvs ln
replaceValue "textarea" (kv:[]) lns  = replaceTextArea kv (unwords lns)
replaceValue "select"   (kv:[]) lns  = replaceSelect kv lns
replaceValue _ _ lns                 = putStr $ unlines lns

subRegex :: String -> String -> String -> String
subRegex str old new = 
  let (a,b,c) = RE.match (RE.makeRegex str :: RE.Regex) old
  in if b=="" then str else a ++ new ++ subRegex c old new

{-- RADIO --}
replaceValueRadio :: KeyValue -> String -> IO ()
replaceValueRadio (k,v) ln = putStrLn newln
    where bareln = subRegex ckd ln "" 
          ckd = "(checked=\"checked\"|checked) *"
          newln = checkRadio v bareln

checkRadio :: String -> String -> String
checkRadio val ln = if vstr `isInfixOf` ln then newln else ln
    where vstr = "value=\"" ++ val ++ "\""
          re = " */>$"
          newln = subRegex re ln " checked=\"checked\" />"

{-- CHECKBOX --}
replaceValueChkbox :: [KeyValue] -> String -> IO ()
replaceValueChkbox kvs ln = putStrLn newln
    where bareln = subRegex ckd ln "" 
          ckd = "(checked=\"checked\"|checked) *"
          newln = checkChkbox kvs bareln

checkChkbox :: [KeyValue] -> String -> String
checkChkbox []          ln = ln
checkChkbox ((k,v):kvs) ln = if f v ln then rep ln else checkChkbox kvs ln
    where f val ln = ("value=\"" ++ val ++ "\"") `isInfixOf` ln
          rep str = subRegex ckd str " checked=\"checked\" />"
          ckd = " */>"

{-- TEXT --}
replaceValueText :: KeyValue -> String -> IO ()
replaceValueText (k,v) ln = putStrLn newln
    where re = "value=\"[^\"]*\""
          str = "value=\"" ++ v ++ "\""
          newln = if isValue ln then subRegex re ln str else insertValue ln str

{-- SELECT --}
replaceSelect :: KeyValue -> [String] -> IO ()
replaceSelect (k,v) lns = putStr $ unlines ans
    where barelns = map (\ln -> subRegex re ln "") lns
          re = " *(selected=\"selected\"|selected) *"
          target = "value=\"" ++ v ++ "\""
          ans = [ if target `isInfixOf` ln then subRegex re2 ln sel else ln | ln <- barelns ]
          re2 = "<option"
          sel = "<option selected=\"selected\""

getSelectBlock :: [String] -> ([String],[String]) -> ([String],[String])
getSelectBlock []       (a,[]) = (a,[])
getSelectBlock (ln:lns) (a,[])
  | "</select>" `isInfixOf` ln = (a ++ [ln],lns)
  | otherwise                  = getSelectBlock lns (a ++ [ln],[])

{-- TEXTAREA --}
replaceTextArea :: KeyValue -> String -> IO ()
replaceTextArea (k,v) ln = putStr $ unlines ans
    where lns = lines $ subRegex re ln ">\n<"
          re = ">.*<"
          re2 = "\\\\n"
          vn = subRegex re2 v "\n"
          ans = (head lns) : vn : (drop 1 lns)

getTextBoxBlock :: [String] -> ([String],[String]) -> ([String],[String])
getTextBoxBlock []       (a,[]) = (a,[])
getTextBoxBlock (ln:lns) (a,[])
  | "</textarea>" `isInfixOf` ln = (a ++ [ln],lns)
  | otherwise                    = getTextBoxBlock lns (a ++ [ln],[])

{-- utilities --}
isValue :: String -> Bool
isValue ln = "value=\"" `isInfixOf` ln

insertValue :: String -> String -> String
insertValue str vstr = unwords $ a ++ [vstr,b]
   where ws = words str
         a = init ws
         b = last ws

checkType :: String -> String
checkType ln
  | "type=\"text\"" `isInfixOf` ln     = "text"
  | "type=\"radio\"" `isInfixOf` ln    = "radio"
  | "type=\"checkbox\"" `isInfixOf` ln = "checkbox"
  | "<textarea " `isInfixOf` ln        = "textarea"
  | "<select " `isInfixOf` ln          = "select"
  | otherwise = "unknown"

cutTargetHtml :: String -> [String] -> ([String],[String])
cutTargetHtml tp []      = ([],[])
cutTargetHtml tp (ln:[]) = ([ln],[])
cutTargetHtml tp lns
  | tp == "textarea"  = getTextBoxBlock lns ([],[])
  | tp == "select"    = getSelectBlock lns ([],[])
  | otherwise         = ([head lns], drop 1 lns)

pickKvs :: [KeyValue] -> String -> [KeyValue]
pickKvs []  _  = []
pickKvs kvs ln = filter (\x -> f x ln) kvs
    where f (k,v) ln = isInfixOf ("name=\"" ++ k ++ "\"") ln

{--
setOpts :: [String] -> Opts
setOpts as = case parse args "" ((unwords as) ++ " ") of
                  Right opt -> opt
                  Left err -> Error ( show err ) 

args = do f1 <- try(filename) <|> return "-"
          char ' '
          f2 <- try(filename) <|> return "-"
          return $ NormalOpts f1 f2

filename = many1 ( try(letter) <|> try(digit) <|> symbol )

symbol = oneOf "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"
--}
